Add sdl3_renderer demo V4

sleeptightAnsiC opened the pull request:
This is yet another PR that implements SDL_renderer demo for libSDL3. I consider this to be a Version 4, because there are 3 other PRs that attempted to do this. This one fixes a lot of long occuring issues and brings some improvements specific to SDL3 API. I've put a lot of time and serious care into this, and I hope it will be useful for someone out there.

PR is split into two logical commits.

EDIT: I collapsed the notes because they were scary to look at, and probably outdated at this point. I doubt anyone fully reads them anyway.

<details><summary>notes [CLICK ME]</summary>
<p>

NOTEs:

- I've started with the code from https://github.com/Immediate-Mode-UI/Nuklear/pull/779 and adjusted it accordingly...
- ...but the full origin is something like: `Nuklear/demo/sdl_renderer` -> https://github.com/Immediate-Mode-UI/Nuklear/pull/772 -> https://github.com/Immediate-Mode-UI/Nuklear/pull/779 + https://github.com/Immediate-Mode-UI/Nuklear/pull/825 -> _this_ PR...
- ...and I can probably put [Co-authored-by](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors#creating-co-authored-commits-on-github) into commit message (just let me let me know who was involved)
- Fixed a ton of issues mentioned by other PRs and reviews. (Sorry... I've stopped tracking those long ago)
- I did NOT bring back `nk_sdl_handle_grab()`, despite that I considered it, due to reasoning from https://github.com/Immediate-Mode-UI/Nuklear/pull/779#discussion_r1966566010
- I did NOT change how `userdata` is handled, despite some concerns, per https://github.com/Immediate-Mode-UI/Nuklear/pull/779/files#r2487912696
- Every `FIXME` in _this_ PR is an issue within Nuklear, that I couldn't solve from the demo side. Each one of those should be already reported and linked to Github. If you see any FIXME without explanation or link, let me know.
- Added "CONFIG" section to main.c which let's you drop any linkage dependency on libm/libc, and use SDL-provided symbols instead (this is currently disabled by default, read FIXME about it)
- Unified the use of malloc/free. Now you can pass one nk_allocator to this demo, and it will be used everywhere when possible. You can also use `nk_sdl_allocator()` to get Nuklear-compatible allocator provided by SDL. This is consistent with how Nuklear API works, but not exactly with the rest of demos.
- Added `nk_sdl_style_set_debug_font()` which let's you use primitive SDL debug font and allows to remove the dependency on font baking entirely (not used by default)
- Made sure that you can use this demo without amalgamated Nuklear header (although this is broken on main branch due to https://github.com/Immediate-Mode-UI/Nuklear/issues/256 )
- Changed CMakeLists.txt to Makefile (because every other demo uses GNU Make, not Cmake)
- I've tested this on: Ubuntu-25.04 with Wayland/Gnome, ArchLinux with Xorg, Windows11 with MSYS2 (although I had some issues due to https://github.com/Immediate-Mode-UI/Nuklear/issues/847#issuecomment-3497050976 ), and on macOS Ventura 13.6.9 - so this is well tested on desktop...
- ...however, I had no chance to test this on mobile platforms (Android/iOS) and on web (Emscripten) - any testing on those will be really helpful
- Suggestions and nitpicks are welcome. I will be force-pushing into this branch as fast as I can.

TODOs (out of the scope of _this_ PR):
- Figure out how to hook [SDL_SetTextInputArea](https://wiki.libsdl.org/SDL3/SDL_SetTextInputArea) into this demo. I haven't found any nk_* functions or any context member that would let me do this.
- I've noticed few issues when using `nk_sdl_style_set_debug_font()`, most of those are related to the Examples being hardcoded to specific font size, but there is also one severe bug: the background of color picker always appears black. Not sure if this is my fault or something in Nuklear.

</p>
</details>

RobLoach commented on demo/common/overview.c:
```diff
@@ -601,6 +601,8 @@ overview(struct nk_context *ctx)
                 nk_flags active;
 
+                nk_labelf(ctx, NK_TEXT_LEFT, "Any Input active: %s", nk_edit_is_any_active(ctx) ? "YES" : "no");
+
```
Was this for testing purposes?

RobLoach commented on demo/sdl3_renderer/main.c:
```diff
@@ -0,0 +1,408 @@
+ * due to symbol resolution issues in some Examples, see:
+ * https://github.com/Immediate-Mode-UI/Nuklear/issues/850
+ * */
+#if 0
```
This would be a follow up issue after this is merged?

sleeptightAnsiC replied:
Since I was already force pushing into this PR, I did exactly what I said in my previous comment.

RobLoach reviewed (commented):
Thanks again for the push one this.

sleeptightAnsiC commented on demo/sdl3_renderer/nuklear_sdl3_renderer.h:
```diff
@@ -0,0 +1,663 @@
+    /* SDL_ConvertEventToRenderCoordinates may mutate 'evt' so we use its copy from now on */
+    NK_MEMCPY(&evtcpy, evt, sizeof(evtcpy));
+
+    if (!SDL_ConvertEventToRenderCoordinates(sdl->renderer, &evtcpy)) {
```
Thanks. Just tested with this and it works fine.

sleeptightAnsiC commented on demo/sdl3_renderer/main.c:
```diff
@@ -0,0 +1,408 @@
+#define NK_VSNPRINTF(s, n, f, a)  SDL_vsnprintf(s, n, f, a)
+/* FIXME: Nuklear ignores NK_STRTOD macro, due to:
+ * https://github.com/Immediate-Mode-UI/Nuklear/issues/851
+ * */
```
I think, I should remove this note. Defining `NK_STRTOD` does not break anything, it's simply being ignored silently. Having a note here is unnecessary and this is already being tracked by https://github.com/Immediate-Mode-UI/Nuklear/issues/851

sleeptightAnsiC replied:
Removed this note note and force pushed.

sleeptightAnsiC commented on demo/sdl3_renderer/main.c:
```diff
@@ -0,0 +1,408 @@
+    #define STBRP_ASSERT          SDL_assert
+    /* There is no need to define STBTT_malloc/STBTT_free macros
+     * Nuklear will define those to user-provided nk_allocator */
+#endif
```
Makes sense.

I removed just the `#ifdef NK_INCLUDE_FONT_BAKING` guard and force pushed it.

sleeptightAnsiC commented on demo/sdl3_renderer/main.c:
```diff
@@ -0,0 +1,408 @@
+            (Uint8)(app->bg.r * 255.0f),
+            (Uint8)(app->bg.g * 255.0f),
+            (Uint8)(app->bg.b * 255.0f),
+            SDL_ALPHA_OPAQUE);
```
Yes. I just changed it to the *Float() one and forced pushed.

I didn't really touch anything related to rendering. It should be pretty much the same as it was back in https://github.com/Immediate-Mode-UI/Nuklear/pull/779 If you see anything else worth changing like this one, let me know.

sleeptightAnsiC commented on demo/sdl3_renderer/nuklear_sdl3_renderer.h:
```diff
@@ -0,0 +1,663 @@
+
+    /* 'text' is an arbitrary place in the buffer where the copy selection starts,
+     * 'len' is NOT the buffer lenght, but instead a count of UTF glyphs
+     * that the selection is trying to copy, we need to count 'buflen' ourselfs */
```
What do you think @RobLoach ? Shall I bring back the old (broken) behavior for the sake of consistency with the rest of the demos? I can just put two implementations here and guard one with `#if 0` guard.

I'm not entirely sure if https://github.com/Immediate-Mode-UI/Nuklear/pull/841 is a correct solution to this problem. I just added a comment there explaining some thoughts: https://github.com/Immediate-Mode-UI/Nuklear/pull/841#issuecomment-3501578904

sleeptightAnsiC replied:
I just added the 2nd implementation next to it and guarder them with preprocess. This is still still using the same behavior, but it will be trivial to change when https://github.com/Immediate-Mode-UI/Nuklear/pull/841 gets merged.

PavelSharp commented on demo/sdl3_renderer/nuklear_sdl3_renderer.h:
```diff
@@ -0,0 +1,663 @@
+    text = SDL_GetClipboardText();
+    NK_ASSERT(text);
+    if (text[0] != '\0')
+        nk_textedit_paste(edit, text, SDL_utf8strlen(text));
```
Please note that all backends based on DirectX/GLFW/SDL tend to pass the number of bytes to `nk_textedit_paste`. Because of that, the paste functionality in each of those backends is currently silently broken, and this is exactly what PR #841 addresses. Passing code point count (as done here) makes the behavior inconsistent and fragile for future changes

sleeptightAnsiC replied:
I just added the 2nd implementation next to it and guarder them with preprocess. This is still using the same behavior, but it will be trivial to change when https://github.com/Immediate-Mode-UI/Nuklear/pull/841 gets merged.

PavelSharp commented:
:tada:Congratulations!
More than 10 years of development and now (aside from a bunch of small bugs) it actually works on Android!
Screenshot 1: running `sdl3_renderer` as-is
Screenshot 2: after adjusting the scale factor(`scale_x`, `scale_y`) and, it fully works at a basic level(navigating menus, interacting with widgets, UI behaves correctly)
What's even better it looks like most of the remaining issues(I guess) aren’t directly related to the `sdl3_renderer` itself, which is very good news. But to be sure, we still need to test it further.
<details>
<summary>CLICK</summary>

<img width="1080" height="1200" alt="изображение" src="https://github.com/user-attachments/assets/ca75416b-411f-4e7e-9725-31ec097e9f22" />


</details>

PavelSharp commented:
@sleeptightAnsiC 

> What's even better it looks like most of the remaining issues(I guess) aren’t directly related to the sdl3_renderer

Honestly, my opinion is still a bit uncertain. As I said, I had to modify these lines:
```
scale_x = 2.5f * (float)(render_w) / (float)(window_w);
scale_y = 2.5f * (float)(render_h) / (float)(window_h);
```
So the main task right now is to understand why the resulting scaling was incorrect. We should lern more about [SDL3 and DPI](https://wiki.libsdl.org/SDL3/README-highdpi)

sleeptightAnsiC commented:
I'll take a look into it.

Callback would be perfect, but I don't want to break ABI/API accidentally. I suspect I simply missed some case for `nk_edit_is_any_active`, I already have some ideas.

When it comes to scale, this code is not mine, it comes from other demos and I didn't dare to touch it. I'll see what can be done here.

I also just realized that I did broke commit history. There should be two commits, not one... I probably smashed those during latest force-push.

PavelSharp commented:
@sleeptightAnsiC 

> When it comes to scale, this code is not mine, it comes from other demos and I didn't dare to touch it. I'll see what can be done here.

Currently, I am working on the High DPI support issue and have the draft solution. Patch will be ready soon

RobLoach commented:
Think it would be okay to merge, and fix the remaining in follow ups? Or should we wait?

sleeptightAnsiC commented:
Right now, I should at least fix the _"keyboard input for all nk\_property\_* is broken"_ https://github.com/Immediate-Mode-UI/Nuklear/pull/852#issuecomment-3507261086, and _"there should be two commits, not one"_ https://github.com/Immediate-Mode-UI/Nuklear/pull/852#issuecomment-350845973, and we are pretty much done, unless something worse pops up. The rest can be patched later.

I don't consider the lack of [SDL_SetTextInputArea](https://wiki.libsdl.org/SDL3/SDL_SetTextInputArea) as a real issue. It's a very advanced feature that everyone will probably want to solve on its own. Only god really knows how this thing is supposed to work :weary: (https://github.com/libsdl-org/SDL/issues/13166).

The problem with DPI scaling that Pavel mentioned (https://github.com/Immediate-Mode-UI/Nuklear/pull/852#issuecomment-3506801190) is also not really an issue. This is something that is solvable by the application. It would be nice to have a good example of this working correctly, but not a must.

@RobLoach In the meantime, would you mind taking a look at `demo/sdl3_renderer/nuklear_sdl3_renderer.h` and other recent changes? I just want to be aware of any concerns ahead of time, so this PR won't take forever to merge.

RobLoach commented:
Looked pretty good to me. Wasn't able to run it since I'll need a CMake implementation, but we can fix any of the other small things up afterwards.

PavelSharp commented:
> I just want to be aware of any concerns ahead of time, so this PR won't take forever to merge.

Just to note. SDL3 has [dropped support for C89](https://github.com/libsdl-org/SDL/issues/6691) in favor of C99/11, so we should definitely keep this in mind.

sleeptightAnsiC commented:
> > I just want to be aware of any concerns ahead of time, so this PR won't take forever to merge.
> 
> Just to note. SDL3 has [dropped support for C89](https://github.com/libsdl-org/SDL/issues/6691) in favor of C99/11, so we should definitely keep this in mind.

That's not exactly true. Every public header should still be in highly portable C (ANSI/C89). They only changed it for internal sources. I don't believe this changes anything for us.

Every major C library uses C89 for public headers, so you can link it with more compilers, but internally they use whatever they want (usually C99/11 with GNU extensions).

(trust me bro, I have "Ansi C" in my name :smile: )

PavelSharp commented:
High DPI support patch is done and ready. 🎉🥳

✅ Key features:
- Runtime / hot DPI scaling based on OS events (e.g. on Windows 10 changing scale to 125% in Settings will immediately trigger UI rebuild and font recalculation).
- Tested on Android 13 as well.
- Comments were added to clarify DPI-related logic and event flow.

I guess, with current Nuklear API there is essentially no better DPI handling possible, we adjust scale via `SDL_SetRenderScale` and rebake fonts on demand. `ProggyClean` won’t remain razor-sharp when scaled (expected limitation), but overall DPI behavior is consistent and correct. So, we have introduced High DPI support for this demo/renderer.
[high_dpi.patch](https://github.com/user-attachments/files/23444974/high_dpi.patch)

rswinkle commented:
I spent a little time today trying to figure out why properties weren't triggering the edit and this is the fix I came up with (not sure if setting the name really matters but I was trying to match code in nk_edit_buffer():

```diff
diff --git a/src/nuklear_property.c b/src/nuklear_property.c
index 68b47ba..a234628 100644
--- a/src/nuklear_property.c
+++ b/src/nuklear_property.c
@@ -405,6 +405,8 @@ nk_property(struct nk_context *ctx, const char *name, struct nk_property_variant
         win->property.name = hash;
         win->property.select_start = *select_begin;
         win->property.select_end = *select_end;
+        win->edit.active = nk_true;
+        win->edit.name = hash;
         if (*state == NK_PROPERTY_DRAG) {
             ctx->input.mouse.grab = nk_true;
             ctx->input.mouse.grabbed = nk_true;
@@ -420,6 +422,7 @@ nk_property(struct nk_context *ctx, const char *name, struct nk_property_variant
         win->property.select_start = 0;
         win->property.select_end = 0;
         win->property.active = 0;
+        win->edit.active = nk_false;
     }
 }
 NK_API void

```
Seems to be working correctly.

I also added a text field to the small demo window to test the differences and discovered something else.  Text input stays active for a text field even if you click on (for instance) a color drop down, even if you actually click around in the color box or drag the nested channel properties. But it goes inactive if it was the property that was being edited. Just another small thing to dig into in the combobox/dropdown code.

sleeptightAnsiC commented:
[replying to multiple messages at the same time]

@rswinkle So this is more of a bug inside of Nuklear, right? If I understand your [comment](https://github.com/Immediate-Mode-UI/Nuklear/pull/852#issuecomment-3518277340), the widgets are not setting their `edit` flags as they should, and this should be fixed "per-widget" (?) I was thinking about re-implementing `nk_edit_is_any_active`, but if the widget context is not properly set, I should probably take a look into this first. 

@PavelSharp I think, I fully understand your [patch](https://github.com/Immediate-Mode-UI/Nuklear/pull/852#issuecomment-3509287687), but I find the font rebaking to be a bit of an overkill. It increases the complexity of `main.c` and this file is meant to be replaced by the demo user anyway. Maybe it would be better to simply set the scale initially with `SDL_GetWindowDisplayScale`, just like you did, and leave the note (and maybe some log message) under `SDL_EVENT_WINDOW_DISPLAY_SCALE_CHANGED` that the font will break if you're not handling it yourself. This is such an edge case that I don't want to add more control flow only for it.

@RobLoach I can bring back the cmake file. I didn't like the old one from https://github.com/Immediate-Mode-UI/Nuklear/pull/779, because it was silently downloading SDL sources instead of reusing the system-wide copy of it. I guess, since SDL3 is now primary cmake-based, it makes sense to have it here.

Anyway, thank you guys for your input and patches. I hope I will deal with this soon. Sorry, for the slowdown on my part. I am a bit exhausted with debugging Nuklear and grepping its old commit messages

rswinkle commented:
> I’d really like to hear what you think about this, @rswinkle and @RobLoach.

I have to agree with @sleeptightAnsiC on this.

> 1. **Avoiding fragile internal changes**
>    Any other approach risks introducing major internal modifications to Nuklear’s core logic, something that could easily lead to regressions and new bugs down the line. Include some complex scenarios that we might not see when testing with the standard examples only
>    A callback, on the other hand, keeps the logic contained and easy to reason about.

I would not call either my approach or the current approach fragile. As sleeptightAnsiC said, all approaches at their optimal/minimal state amount to the same thing, and in fact the callback method probably requires the most in actual lines changed. The logic is not/would not be in the callback. It is a matter of Nuklear internal state regardless. Knowing where/when to insert the callback is the same knowledge you would need for any other method.

> 2. **Currently, there are already many bugs and fixing them may lead to new bugs**

Again, these are Nuklear bugs that need to be fixed in any case and the code changes are very small. It's a matter of discovery/precision.

> 3. **Respecting user-defined widgets**
>    A callback also gives us a clean and extensible way for custom widgets (not based on `nk_do_edit`) to request and release text input properly, something impossible to do safely under the current proposal.

The request and release does not happen at the widget level, that's part of the issue. It has to be aggregated over all the windows/widgets over the whole frame being rendered, and at that point, you know if either Start or Stop should be called based on whether it changed from the previous. Whether you use my method, adding a couple of members to try to simplify the test by collating as we go, or try to get the same result using a function like nk_is_any_active() combining hopefully all necessary checks (or both), the solution is the same.  Ideally all widgets using text input should use the same method directly or (since that's obviously not the case) they should pass that state along to the text_edit state as the "master" state.

> 4. **Safe and backward-compatible**
>    If implemented with a macro guard, this feature remains completely optional and non-breaking.
>    Existing applications would behave exactly the same, while backends or advanced users could enable the callback when needed.

There is no breaking change regardless.  Nuklear is not used as a dynamic library that will be updated separately from the application(s) that use it. It is meant to be compiled statically. And the changes are not what I would consider "breaking": Either a couple added field to an internal structure, an extra utility function meant almost exclusively for backend implementers not average users, or both.  No change to the existing API, and nothing removed that could break someone's code.

> 5. **Contextual data**(mostly for `SDL_SetTextInputArea`)

I still don't really understand the point of this function but all the logic above applies here too afaict. The difficulty is in getting the information in the first place.

> 6. Consistency with existing Nuklear design
>    This is just my opinion, but adding a virtual keyboard callback is not a hacky workaround, it’s actually a clean and natural extension of Nuklear’s design. It fits perfectly into the existing family of callbacks (copy, paste, etc.) and keeps the logic modular, explicit, and easy to reason about, exactly in the spirit of Nuklear’s minimalistic philosophy.

None of them are hacky workarounds. As I said all would require roughly the same logic/changes, the callback would just require a bit more code for the macro stuff at the top. My first attempt at doing this was a macro based callback system a la NK_IS_WORD_BOUNDARY() which I worked on a while back. It was two optional macros NK_ON_EDIT_FOCUS()/NK_ON_EDIT_UNFOCUS().  I ended up deleting it and replacing it twice, ending up on my current version because I realized everything that I discussed above. Not only do we not want or need to call Start/StopText() within Nuklear (even indirectly), but everything boils down to correctly and efficiently tracking the boolean state so the backend/user can do it themselves if they want to. 

What should be documented is all the knowledge about how that state is tracked which is what someone implementing a new text input widget not based on the nk_do_edit() family needs to know. As long as they remember the few lines to set the state appropriately then everything would just work.


>It’s also important to note that nk_do_edit is only called by text fields (e.g. nk_edit_string) and nk_property. This means the callback would directly address the exact issue we are discussing, without affecting unrelated parts of Nuklear.
>I believe I’ve given enough reasons why a new callback is preferable. Modifying nk_edit_is_any_active would require many changes in Nuklear, which almost equivalent to new bugs.

See above. It's not that many changes and you would have to touch the same pieces of the code for the callback method.

sleeptightAnsiC commented on src/nuklear_edit.c:
```diff
@@ -833,4 +833,12 @@ nk_edit_string_zero_terminated(struct nk_context *ctx, nk_flags flags,
+    if (!ctx) return nk_false;
 
+    return (ctx->active && ctx->active->edit.active) ||
+           (ctx->active && ctx->active->popup.win && ctx->active->popup.win->edit.active);
```
> I also added a text field to the small demo window to test the differences and discovered something else. Text input stays active for a text field even if you click on (for instance) a color drop down, even if you actually click around in the color box or drag the nested channel properties. But it goes inactive if it was the property that was being edited. Just another small thing to dig into in the combobox/dropdown code.

Reproduced! and I'm pretty sure there is a bug on this line, because combobox/colorpicker is actually a "popup" under the hood. It is definitely wrong to consider the edit from the window active, while the popup without active edit is open.

<img width="425" height="516" alt="image" src="https://github.com/user-attachments/assets/2cee305a-30ad-45b6-90cd-123f60847619" />

sleeptightAnsiC replied:
Replying to https://github.com/Immediate-Mode-UI/Nuklear/pull/852#issuecomment-3518639254

> Similarly, I'm sure when I look at the combobox/dropdown code will have its own state and will be missing the lines to turn off an active edit. I'll do that later if someone else hasn't already by then.

I'm pretty sure, this was the same issue.

sleeptightAnsiC replied:
fixed with latest change

sleeptightAnsiC commented:
Replaying to https://github.com/Immediate-Mode-UI/Nuklear/pull/852#issuecomment-3518277340

> not sure if setting the name really matters but I was trying to match code in nk_edit_buffer

This is generally desired and probably every widget should do this, however, I just tested this and some are not setting/clearing the edit.name hash, which complicates my life a lot. Hah (I'm crying inside)

---

Replaying to https://github.com/Immediate-Mode-UI/Nuklear/pull/852#issuecomment-3519350868

> I'm planning to change nk_edit_is_any_active to nk_edit_get_active
> [...]
> I don't remember weather global contex had a reference to the edit widget (I hope it has).

Welp, it does not. It only has the `edit.name` hash and this doesn't really help, because this hash is only useful for whoever set it in the first place. We could try to compare the `window->edit.name` with something like `window->property.name` and do some magic accordingly, but that would be short-sighted and fragile.

---

At this point, I think I can add two fields to `nk_edit_state`, one for edit bounds and one for cursor bounds, and I can try hooking them somewhere inside of `nk_do_edit`. I'm not entirely sure if that would be a right decision. I've taken a look at what Pavel tried to do in https://github.com/Immediate-Mode-UI/Nuklear/pull/857 and he generally had a similar idea, but chosen `nk_keyboard` instead. Me attempting to fix it here could make his changes from https://github.com/Immediate-Mode-UI/Nuklear/pull/857 incompatible, so that's another thing I'm trying to avoid.

PavelSharp commented:
> Your solution from https://github.com/Immediate-Mode-UI/Nuklear/pull/857 also triggers the callbacks from nk_do_edit, so how the "not based on nk_do_edit" case is being solved there? What do you mean by "user-defined widgets which are not based on nk_do_edit" ?

Let's imagine a user-defined widget that wants to notify Nuklear's input system (and backend), that it is about to receive text keyboard input.

In this case, the widget is responsible for calling `nk_input_want_text_keyboard`, and if that succeeds, it must populate the `text_cursor` and `text_area` fields. In other words, the responsibility for requesting text keyboard input and filling these fields lies entirely with the widget.

A clear example of this is `nk_do_edit`, which already handles this, but any other widget could do the same.

The main point is that `text_want`, `text_area`, and `text_cursor` belong to the input system’s notification mechanism - a new concept which was introduced by my PR. We update `nk_keyboard` with these fields, to form the input notification system. Since the backend checks them before nk_begin, no callbacks are needed.

sleeptightAnsiC commented:
I've taken a better look at https://github.com/Immediate-Mode-UI/Nuklear/pull/857 and I'm a bit confused.

> the widget is responsible for calling nk_input_want_text_keyboard, and if that succeeds, it must populate the text_cursor and text_area fields [...] a new concept which was introduced by my PR

Why would people follow this very specific _"concept"_, if they're not even following an already established solution like `nk_do_edit()` ? This reasoning contradicts itself and can be applied the other way around.

> We update nk_keyboard with these fields, to form the input notification system. Since the backend checks them before nk_input_begin, no callbacks are needed.

This is not a notification, no one is getting notified about anything. The user of https://github.com/Immediate-Mode-UI/Nuklear/pull/857 would still have to pull that information manually from `ctx->input.keyboard` (or similar) which also expects that the widget must set and update it correctly.

This really feels like repeating the old mistake. We already proven that widgets are not to be trusted with setting the context state correctly, and against all reasoning, this new "concept" introduces yet another state that we have to track and care about. It does not create any convention that custom widgets could safely follow.

I don't get it. All your previous messages were about a _"callback"_ system (I see 43 matches when I search for this word on this page). You've tried very hard to convince us to _"calback"_ solution, so why this new thing ended up being another manual _"state"_ update? I'm really confused here. Note that [I only disagreed about callbacks being a part of _this_ PR](https://github.com/Immediate-Mode-UI/Nuklear/pull/852#issuecomment-3519049014), I still thought you would go for it later in another PR (https://github.com/Immediate-Mode-UI/Nuklear/pull/857).

PavelSharp commented:
> This is not a notification, no one is getting notified about anything. The user of https://github.com/Immediate-Mode-UI/Nuklear/pull/857 would still have to pull that information manually from ctx->input.keyboard (or similar)

That’s not quite right. The backend is the one that needs to read this information, the user doesn’t have to worry about it.
I only used the word “notification” to help explain the idea. I often work with publisher-subscriber patterns, so the term just felt natural to me in this context.

> also expects that the widget must set and update it correctly.

Yes, if a user wants to create a custom, non-standard text-input widget, they should use this new simple API. In this very rare situation, this behavior is completely normal, and it has already been mentioned here (quoting it for the second time).

> What should be documented is all the knowledge about how that state is tracked which is what someone implementing a new text input widget not based on the nk_do_edit() family needs to know. As long as they remember the few lines to set the state appropriately then everything would just work.

**Next**

> This really feels like repeating the old mistake. We already proven that widgets are not to be trusted with setting the context state correctly, and against all reasoning

It’s possible that the earlier arguments weren’t convincing, because #857 works very well and very stable. I haven’t been able to find any issues, even when testing different combinations of clicks and focus changes. 
Also, #857 results in only about 30 lines of final changes and it’s important to recognize that these 30 lines are the conclusion of a long and difficult effort to properly integrate SDL3.

> this new "concept" introduces yet another state that we have to track and care about. It does not create any convention that custom widgets could safely follow.

**First**, this new state exists only until the next call to `nk_input_begin`, which normally corresponds to a single frame. Because of that, introducing this state is harmless. 

**Second**, please read my quote again:

> calling nk_input_want_text_keyboard, and if that succeeds, it must populate the text_cursor and text_area fields.

So, the rule is very simple.

> I don't get it. All your previous messages were about a "callback" system (I see 43 matches when I search for this word on this page). You've tried very hard to convince us to "calback" solution, so why this new thing ended up being another manual "state" update? I'm really confused here. 

In the end, my opinion changed. You may want to look more closely at #857, where it is explicitly mentioned:

> The chosen design is [inspired by Dear ImGui](https://github.com/ocornut/imgui/blob/c254db763709e12eeabea54c52631c2b972b7d58/backends/imgui_impl_sdl3.cpp#L178-L207). ImGui demonstrates that callback-based control can be unnecessarily heavy

**as well as:**

> Since the backend checks them before `nk_input_begin`, no callbacks are needed.

sleeptightAnsiC committed:
Fix(nk_property): property widget not updating 'win->edit.active' state

Other "editable" widgets are updating the its window 'edit.active' state,
but not nk_property. SDL3 demo needs this in order to hook SDL_SetTextInput.
This patch was initially proposed by https://github.com/rswinkle
The only thing I changed in it, was a removal of 'edit.name' update,
nk_property is not using the name hash, so there is no reason to set it.

Origin: https://github.com/Immediate-Mode-UI/Nuklear/pull/852#issuecomment-3518277340
Co-authored-by: Robert Winkler <rob121618@gmail.com>

sleeptightAnsiC committed:
Add new demo: sdl3_renderer

This is a highly updated demo for libSDL3. It fixes a lot of long
occuring issues and brings some improvements specific to SDL3 API.
The origin and development of this demo is quite long.
The exact track of development is something like this:
`Nuklear/demo/sdl_renderer` -> https://github.com/Immediate-Mode-UI/Nuklear/pull/772 -> https://github.com/Immediate-Mode-UI/Nuklear/pull/779 + https://github.com/Immediate-Mode-UI/Nuklear/pull/825 -> _this_ commit...
You may wish to read the linked PRs in order to get the full context.
Commit message is too small to describe everything. I'm sorry...

sleeptightAnsiC commented:
The latest force-push brings following changes:
- Fix for initial font scaling, and added a note about how to deal with potential rescaling at runtime
- Fix for nk_property not updating 'edit.active' state
- Added a check that triggers when something else than demo is trying to SDL_SetTextInputActive
- Moved the SDL_ConvertEventToRenderCoordinates from `nk_sdl_handle_event` to the `main.c` layer (I think the developer should be handling this)
- Given the latest discoveries about how Nuklear Widgets handle its states, I...
- ...nuked `nk_edit_is_any_active()` and moved its functionality to demo to `nk_sdl_update_TextInput()`
- ...nuked the debug label from `overview.c` and moved its functionality to demo to `main.c`
- ...introduced new demo function `nk_sdl_update_TextInput()` and left there a FIXME about possible SDL_SetTextInputArea implementation
- PR is again split into two separate commits, so I converted it from Draft

I hope, this will bring a compromise to how TextInput should be tracked and updated, at least for now. Nothing changes in nk_* API any more. I made it so it will be easy to tweak TextInputActive and solve TextInputArea in the future. I may have some disagreement with Pavel, but I don't want to sabotage his work from https://github.com/Immediate-Mode-UI/Nuklear/pull/857 This is the most "friendly" solution I can see at the moment, and I hope that everyone will understand.

Please, do not drag this PR for any longer with features that can be solved later.
Let's retest this thing one last time, merge it, and move on to other problems.
There is so much I want to upstream and this demo is taking all my attention.

@PavelSharp Maybe, let's hide our latest comments, and/or move their content to https://github.com/Immediate-Mode-UI/Nuklear/pull/857 ?

PavelSharp commented:
@sleeptightAnsiC 

> Maybe, let's hide our latest comments, and/or move their content to #857?

Yes, I agree. This discussion relates much more to #857. So let's hide our messages here, and I'll move it.

> I may have some disagreement with Pavel, but I don't want to sabotage his work from https://github.com/Immediate-Mode-UI/Nuklear/pull/857

Actually, I'm open to your suggestions. If you have _specific_ ideas about what changes we should make in #857, please feel free to share them. I appreciate it

zoogies commented:
Woo! Thanks, all. Glad to see this merged! Shoutout to @sleeptightAnsiC for taking the lead on this
